src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java

```diff
@@ -0,0 +1,253 @@
+    // Run at batch priority, which won't count toward concurrent rate limit.
+    if (config.getMode().equals(QueryJobConfiguration.Priority.BATCH)) {
+      builder.setPriority(QueryJobConfiguration.Priority.BATCH);
+    } else {
+      builder.setPriority(QueryJobConfiguration.Priority.INTERACTIVE);
+    }
+
+    // Save the results of the query to a permanent table.
+    if (config.getDataset() != null && config.getTable() != null) {
+      builder.setDestinationTable(TableId.of(config.getDataset(), config.getTable()));
+    }
+
+    // Enable or Disable the query cache to force live query evaluation.
+    if (config.shouldUseCache()) {
+      builder.setUseQueryCache(true);
+    }
+
+    // Enable legacy SQL
+    builder.setUseLegacySql(config.isLegacySQL());
+
+    QueryJobConfiguration queryConfig = builder.build();
+
+    // Location must match that of the dataset(s) referenced in the query.
+    JobId jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+
+    // API request - starts the query.
+    Credentials credentials = config.getServiceAccountFilePath() == null ?
+                                null : GCPUtils.loadServiceAccountCredentials(config.getServiceAccountFilePath());
+    BigQuery bigQuery = GCPUtils.getBigQuery(config.getProject(), credentials);
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
+    LOG.info("Executing SQL as job {}.", jobId.getJob());
+    LOG.debug("The BigQuery SQL is {}", config.getSql());
+
+    // Wait for the query to complete
+    queryJob.waitFor();
+
+    // Check for errors
+    if (queryJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+    }
+
+    TableResult queryResults = queryJob.getQueryResults();
+    long rows = queryResults.getTotalRows();
+    if (config.shouldSetAsArguments()) {
+      if (rows == 0 || queryResults.getSchema() == null) {
+        LOG.warn("The query result does not contain any row or schema, will not save the results in the arguments");
+      } else {
+        Schema schema = queryResults.getSchema();
+        FieldValueList firstRow = queryResults.iterateAll().iterator().next();
+        for (int i = 0; i < schema.getFields().size(); i++) {
+          Field field = schema.getFields().get(i);
+          String name = field.getName();
+          if (field.getType().equals(LegacySQLTypeName.RECORD)) {
+            LOG.warn("Field {} is a record type with nested schema, will not save the value in the argument", name);
+            continue;
+          }
+          context.getArguments().set(name, firstRow.get(name).getStringValue());
+        }
+      }
+    }
+
+    context.getMetrics().gauge(RECORDS_PROCESSED, rows);
```

Comment:
use context.getStageMetrics(), which will automatically prefix the stage name. Otherwise this will clobber metrics from other stages in the pipeline.

Reply:
`context.getMetrics` already returns a `StageMetrics`, which I believe will prefix with stage name

Reply:
ah I see, nevermind then